Eng/pr 85476351

cachemeifyoucan opened the pull request:
Dont merge yet. Looking for feedback for some of the changes since I don't know if they are good enough or not.

The exclude path is to remove `llvm-jitlink` directory because the object files built there has section groups which are not supported by current ELF linkgraph builder yet.

cachemeifyoucan committed:
add normalization for ELF TLS EdgeKind

cachemeifyoucan committed:
[llvm-cas-object-format] Add an option to exclude path from ingestion

Add an option to exclude paths from CAS so we can skip files that is not
currently supported by JITLink.

cachemeifyoucan commented:
Some data for ingesting llvm-project ELF from this branch:
Debug:
```
Kind                        Count            Parents           Children           Data (B)           Cost (B)
====                        =====            =======           ========           ========           ========
builtin:blob                    1   0.00%          8   0.00%          0   0.00%         21   0.00%         21   0.00%
builtin:node                    8   0.00%       2425   0.03%         15   0.00%        114   0.00%        414   0.00%
builtin:tree                  690   0.03%        689   0.01%       3119   0.04%     105820   0.02%     168200   0.03%
cas.o:block                206435   9.29%    1253054  16.84%          0   0.00%  152586699  34.63%  152586699  25.89%
cas.o:compile-unit           2418   0.11%       2430   0.03%    7437894  99.96%   77736342  17.64%  226494222  38.43%
cas.o:section              481800  21.69%    1823947  24.51%          0   0.00%   55794025  12.66%   55794025   9.47%
cas.o:symbol              1530112  68.88%    4358475  58.57%          0   0.00%  154346724  35.03%  154346724  26.19%
TOTAL                     2221464 100.00%    7441028 100.00%    7441028 100.00%  440569745 100.00%  589390305 100.00%

Kind                        Count            Parents           Children           Data (B)           Cost (B)
====                        =====            =======           ========           ========           ========
builtin:blob                 2428  77.72%       2430  77.76%          0   0.00% 1175712720  99.99% 1175712720  99.99%
builtin:tree                  696  22.28%        695  22.24%       3129 100.00%     106152   0.01%     168732   0.01%
TOTAL                        3124 100.00%       3125 100.00%       3129 100.00% 1175818872 100.00% 1175881452 100.00%
```

Release
```
Kind                        Count            Parents           Children           Data (B)           Cost (B)
====                        =====            =======           ========           ========           ========
builtin:blob                    1   0.00%          8   0.00%          0   0.00%         21   0.00%         21   0.00%
builtin:node                    8   0.00%       2547   0.26%         15   0.00%        114   0.00%        414   0.00%
builtin:tree                  783   0.14%        783   0.08%       3338   0.34%     112488   0.08%     179248   0.12%
cas.o:block                120651  22.22%     236291  23.85%          0   0.00%   93873847  69.07%   93873847  60.28%
cas.o:compile-unit           2540   0.47%       2555   0.26%     987384  99.66%    8110778   5.97%   27858458  17.89%
cas.o:section              111934  20.61%     178354  18.00%          0   0.00%   10667303   7.85%   10667303   6.85%
cas.o:symbol               307147  56.56%     570199  57.55%          0   0.00%   23155631  17.04%   23155631  14.87%
TOTAL                      543064 100.00%     990737 100.00%     990737 100.00%  135920182 100.00%  155734922 100.00%

Kind                        Count            Parents           Children           Data (B)           Cost (B)
====                        =====            =======           ========           ========           ========
builtin:blob                 2553  76.39%       2555  76.41%          0   0.00%  221526992  99.95%  221526992  99.92%
builtin:tree                  789  23.61%        789  23.59%       3348 100.00%     112820   0.05%     179780   0.08%
TOTAL                        3342 100.00%       3344 100.00%       3348 100.00%  221639812 100.00%  221706772 100.00%
```

cachemeifyoucan commented:
I will split some of the commits to go upstream. Current status is enough to convert llvm-project build.

cachemeifyoucan commented:
https://reviews.llvm.org/D114747
https://reviews.llvm.org/D114749
https://reviews.llvm.org/D114753

cachemeifyoucan committed:
[JITLink][ELF] Don't skip sections of size 0

Size 0 sections can have symbols that have size 0. Build those sections
and symbols into the LinkGraph so they can be used properly if needed.

Reviewed By: lhames

Differential Revision: https://reviews.llvm.org/D114749

cachemeifyoucan committed:
[JITLink][ELF] Add support for reading extended table

Add support for reading extended table in ELF object file. This allows
JITLink to support ELF object files with many sections.

Reviewed By: lhames

Differential Revision: https://reviews.llvm.org/D114747

dexonsmith commented:
Happy for this to land at any point; maybe easier to finish reviewing upstream and then cherry-pick. Might make sense to land the `llvm-cas-object-format` right away (that change LGTM to push on its own).

cachemeifyoucan committed:
[JITLink][ELF] Support duplicated section names from object file

ELF object files can contain duplicated sections (thus section symbols
as well), espeically when comdats/section groups are present. This patch
adds support for generating LinkGraph from object files that have
duplicated section names. This is the first step to properly model
comdats/section groups.

Differential Revision: https://reviews.llvm.org/D114753
